joint_state_publisher/joint_state_publisher/joint_state_publisher.py

```diff
@@ -187,33 +220,39 @@ def init_urdf(self, robot):
 
             if jtype == 'continuous':
                 joint['continuous'] = True
-            self.free_joints[name] = joint
+            free_joints[name] = joint
+
+        return (free_joints, joint_list, dependent_joints)
 
     def configure_robot(self, description):
         self.get_logger().debug('Got description, configuring robot')
         try:
-            robot = xml.dom.minidom.parseString(description)
+            xmldom = xml.dom.minidom.parseString(description)
         except xml.parsers.expat.ExpatError:
             # If the description fails to parse for some reason, print an error
             # and get out of here without doing further work.  If we were
             # already running with a description, we'll continue running with
             # that older one.
-            self.get_logger().warn('Invalid robot_description given, ignoring')
+            self.get_logger().warn('Invalid robot description given, ignoring')
+            return
+
+        root = xmldom.documentElement
+        try:
+            if root.tagName == 'sdf':
+                (free_joints, joint_list, dependent_joints) = self.init_sdf(xmldom)
+            elif root.tagName == 'COLLADA':
+                (free_joints, joint_list, dependent_joints) = self.init_collada(xmldom)
+            else:
+                (free_joints, joint_list, dependent_joints) = self.init_urdf(xmldom)
+        except Exception as e:
+            self.get_logger().warn('Invalid robot description: %s' % (str(e)))
```

Comment:
What kind of exceptions can we get here?

Reply:
Good question.

It looks to me like there are a few different places that could possibly throw exceptions.  For instance, if the SDF file was missing the `model` element at line https://github.com/ros/joint_state_publisher/blob/0ea02939f79899cf1cb51edeff3d09bd62553afb/joint_state_publisher/joint_state_publisher/joint_state_publisher.py#L70 , we would get an `IndexError` exception.  Similarly, `init_collada` can throw exceptions if the COLLADA file is of the wrong version, or if there is no `kinematics_model` or `technique_common` included in the XML (https://github.com/ros/joint_state_publisher/blob/0ea02939f79899cf1cb51edeff3d09bd62553afb/joint_state_publisher/joint_state_publisher/joint_state_publisher.py#L122-L130).  Overall, anytime we have a malformed XML file (of whatever type), I want to make sure that we ignore it.

Actually, though, that brings up something I hadn't considered before.  There are actually two situations in which `configure_robot` is called.  The first situation is waiting for a valid description on the `robot_description` topic during runtime.  At that time, we definitely don't want to crash if we get an invalid XML.  The second situation where this is called is when the robot description is passed on the command-line.  If we fail parsing there, then we do actually want to quit because we can't do much meaningful, and it is better to warn the user early on.

Because of that, I think I'm going to change `configure_robot` to return a boolean, and check it only in the case it was passed on the command-line.

Reply:
See 16d124666bce8dcd7991a14ef5e96218e041c574